# COMPUTACIÓN
## Práctica 2
### Rodrigo Segura Moreno
_____

Descomprimimos la base de datos:

In [1]:
tar -xzf base.tar.gz

Creamos un directorio llamado __Resultados__

In [2]:
mkdir Resultado

___
## cat,  head, tail

__1.__ Copia el archivo (utilizando comandos de bash) que se encuentra en `directorio/TLALPAN/sectorH/names_2009-08-30.txt` a la carpeta `Resultado`.  

Ahora colócate en el directorio `Resultado` y ejecuta los siguientes comandos, para responder a las siguientes preguntas:  
    

In [3]:
cat directorio/TLALPAN/sectorH/names_2009-08-30.txt > Resultados/names_2009-08-30.txt

bash: Resultados/names_2009-08-30.txt: No such file or directory



___
__2.__ ¿Cuál es el resultado de ejecutar los siguientes comandos en la terminal? 

    a) `cat < names_2009-08-30.txt > names_cat.txt`
    
    b) `cat < names_2009-08-30.txt >> names_cat.txt`

In [4]:
pwd
cd Resultado

/home/jovyan/SistemaOperativo_LINUX/Practica_2



    a) El comando siguiente copia la información del archivo *names_2009-08-30.txt* a un nuevo documento *names_cat.txt*

In [5]:
cat < names_2009-08-30.txt > names_cat.txt

bash: names_2009-08-30.txt: No such file or directory



    b) El comando siguiente concatena la información del archivo *names_2009-08-30.txt* a al archivo *names_cat.txt*

In [6]:
cat < names_2009-08-30.txt >> names_cat.txt

bash: names_2009-08-30.txt: No such file or directory



___
__3.__ ¿Cuál es la diferencia entre ambos? Explícalo con tus propias palabras. 

La diferencia entre el comando "__cat__  _archivo1_ >  _archivo2_" y el comando "__cat__  _archivo1_ >>  _archivo2_" radica en que, el primer comando copia la información del _archivo1_ al _archivo2_ sobreescribiendo lo que sea que este en el _archivo2_. El segundo comando concatena la información de _archivo1_ al final del _archivo2_.

___
__4.__ ¿Cuál es la diferencia entre estos comandos?
 
    a) `cat names.txt 2> err.dat`
    
    b) `cata names.txt 2> err.dat`
    

__La__ diferencia entre el comando __a)__ y el comando __b)__ es que, el comando __cat__ si existe y el comando __cata__ no es un comando existente (al menos no lo es en Binder)

___
__5.__ Por último ejecuta: `cat > feliz` enseguida escribe algo y luego ejecuta `Ctrl+D` dos veces. ¿qué fue lo que sucedio?

Se genera un archivo __feliz__ dentro de la carpeta de _Resultados_ el cual contiene el texto que introdujimos.

___
__6.__ ¿Existe algúna forma, con los comandos de bash, de mostrar el archivo de forma inversa? En caso de existir, utilízalo para ver el archivo `directorio/COYOACAN/sectorG/names_2005-09-27.dat` y guarda la salida en el archivo `catinverso.dat` 

Sí, la forma más sencilla es usando el comando __tac__.

In [7]:
tac ../directorio/COYOACAN/sectorG/names_2005-09-27.dat > catinverso.dat

In [8]:
head catinverso.dat


Lopez Martin	5270565437

Penilla Emmanuel	5990448441

Catalan Martha Mariana	5274177518

Arredondo Jorge Alberto	5944048778

Galindo Rodrigo	5292735418



___
__7.__ Crea un solo archivo con el nombre _todosarchivos.txt_ que tenga el contenido de todos los archivos del directorio `directorio/CUAUHTEMOC/sectorE/` 

In [9]:
cat ../directorio/CUAUHTEMOC/sectorE/* >> todosarchivos.txt

Fácilmente podemos corroborar si lo que se hizo fue correcto dirigiéndonos a la ruta `directorio/CUAUHTEMOC/sectorE/` y se verifican las primeras líneas de primer archivo y las últimas líneas del último archivo. Una vez hecho esto podemos hacer lo mismo para el archivo `Resultados/todosarchivos.txt` y corroborar si el procedimiento se hizo correctamente.

Primeras líneas del primer archivo en `directorio/CUAUHTEMOC/sectorE/`

In [10]:
head ../directorio/CUAUHTEMOC/sectorE/names_2001-01-14.dat

Razo Pedro	5078328708

Zavala Oscar Enrique	5349803565

Arevalo Anabel	5098283547

Castillo Regina	5585642291

Muñoz Alexia	5670317235




Últimas líneas del último archivo en `directorio/CUAUHTEMOC/sectorE/`

In [11]:
tail ../directorio/CUAUHTEMOC/sectorE/names_2020-10-17.dat

Prieto Ariana Joselyn	5503241630

Hernández Melanie Daniela	5640530039

Benitez Leonardo	5188101544

Torres Karla Mariell	5267975011

Solís Oscar Eduardo	5808633499




Primeras líneas del archivo `Resultados/todosarchivos.txt`

In [12]:
head todosarchivos.txt

Razo Pedro	5078328708

Zavala Oscar Enrique	5349803565

Arevalo Anabel	5098283547

Castillo Regina	5585642291

Muñoz Alexia	5670317235




In [13]:
tail todosarchivos.txt

Prieto Ariana Joselyn	5503241630

Hernández Melanie Daniela	5640530039

Benitez Leonardo	5188101544

Torres Karla Mariell	5267975011

Solís Oscar Eduardo	5808633499




Así pues, podemos notar que las líneas coinciden. Podemos entonces asumir que el procedimiento fue el correcto.

___
__8.__ Coloca en un archivo con el nombre `colacabeza` las primeras 23 líneas numeradas y las últimas 32 líneas numeradas del archivo `todosarchivos.txt`

*Opcional*: Numera las líneas.  Genera un archivo donde se indiquen con numeros las líneas de este archivo.

Primeramente, se enumerará todo el archivo `todosarchivos.txt` y generaremos un archivo `borrrame.txt` con esa información. Posteriormente vamos a sobreescribir la información en `colacabeza.txt` con sus primeras 23 líneas y sus últimas 32.

In [14]:
# Tenemos que saber cuántas líneas hay en el archivo para proceder
wc -l  todosarchivos.txt

800 todosarchivos.txt



In [15]:
head -800 todosarchivos.txt | nl -b a > borrame.txt

In [16]:
head -24 borrame.txt > colacabeza.txt # Contamos 24 para poder añadir una línea en blanco como un "espacio"
tail -32 borrame.txt >> colacabeza.txt

In [17]:
# Se elimina borrame.txt, ya no es necesario
rm borrame.txt

In [18]:
# Se muestra el archivo colacabeza.txt
cat colacabeza.txt

     1	Razo Pedro	5078328708
     2	
     3	Zavala Oscar Enrique	5349803565
     4	
     5	Arevalo Anabel	5098283547
     6	
     7	Castillo Regina	5585642291
     8	
     9	Muñoz Alexia	5670317235
    10	
    11	Molina Montserrat	5506646832
    12	
    13	Paredes Jose Manuel	5683845245
    14	
    15	Meza Jovanna Sofia	5197111714
    16	
    17	Hernandez Ana Karen	5000717497
    18	
    19	Cejudo Elsy Daniela	5571564174
    20	
    21	Sarabia Ana Ximena	5563916832
    22	
    23	Cerezo Santiago Miguel	5018574468
    24	
   769	Maya Ximena Itzel	5187140592
   770	
   771	Cazares Heriberto	5591260956
   772	
   773	Saldaña Katia Mariana	5297695281
   774	
   775	Balandra Mabel	5346871896
   776	
   777	Aguilar Maria Fernanda	5620980938
   778	
   779	Cazares Ximena	5031790603
   780	
   781	Nieto Oscar Enrique	5529321728
   782	
   783	Ruiz Alexis	5330479468
   784	
   785	Serrano Luis Fernando	5429722585
   786	
   787	Avila Monserrat	5896435866
   788	
   789	Ramos Martha	5849299706
 

___
## Permisos

__9.__ Explica que hace el comando `umask` y pon un ejemplo de su uso. 

In [19]:
umask --help

umask: umask [-p] [-S] [mode]
    Display or set file mode mask.
    
    Sets the user file-creation mask to MODE.  If MODE is omitted, prints
    the current value of the mask.
    
    If MODE begins with a digit, it is interpreted as an octal number;
    otherwise it is a symbolic mode string like that accepted by chmod(1).
    
    Options:
      -p	if MODE is omitted, output in a form that may be reused as input
      -S	makes the output symbolic; otherwise an octal number is output
    
    Exit Status:
    Returns success unless MODE is invalid or an invalid option is given.



Como sugiere la ayuda de dicho comando, `umask` asinga permisos de archivo predeterminados para carpetas y archivos recién creados.

___
__10.__ Copia la carpeta `CUAUHTEMOC` dentro de `Resultados/` y cambia los permisos de todos los archivos de la siguiente manera sólo lectura y escritura para el dueño, sólo ejecución para el grupo y sólo lectura para los demás.

In [20]:
# Creamos el directorio
mkdir -p CUAUHTEMOC
# Generamos la copia dentro del directorio
cp -R ../directorio/CUAUHTEMOC/* CUAUHTEMOC/

In [21]:
# Quitamos todos los permisos a todos para asegurarnos para que al asignar permisos sean exactamento los que queremos
chmod a-rwx CUAUHTEMOC/*/*

In [22]:
# Asignamos los permisos correspondientes
chmod u+rw CUAUHTEMOC/*/*
chmod g+x CUAUHTEMOC/*/*
chmod o+r CUAUHTEMOC/*/*

Revisaremos los permisos de los primeros y los últimos archivos de la carpeta para corroborar que sí se le asignaron los permisos de forma correcta.

In [23]:
# Primeros archivos
ls -l CUAUHTEMOC/* | head -10

CUAUHTEMOC/sectorA:
total 80
-rw---xr-- 1 jovyan root 618 Nov  7 06:37 names_2000-07-13.txt
-rw---xr-- 1 jovyan root 627 Nov  7 06:37 names_2001-04-24.txt
-rw---xr-- 1 jovyan root 634 Nov  7 06:37 names_2002-12-26.dat
-rw---xr-- 1 jovyan root 583 Nov  7 06:37 names_2004-01-19.dat
-rw---xr-- 1 jovyan root 671 Nov  7 06:37 names_2006-03-04.dat
-rw---xr-- 1 jovyan root 617 Nov  7 06:37 names_2006-06-28.dat
-rw---xr-- 1 jovyan root 589 Nov  7 06:37 names_2006-07-15.txt
-rw---xr-- 1 jovyan root 613 Nov  7 06:37 names_2006-08-22.txt
ls: write error: Broken pipe



In [24]:
# Últimos archivos
ls -l CUAUHTEMOC/* | tail -10

-rw---xr-- 1 jovyan root 613 Nov  7 06:37 names_2009-03-12.dat
-rw---xr-- 1 jovyan root 622 Nov  7 06:37 names_2009-04-20.dat
-rw---xr-- 1 jovyan root 631 Nov  7 06:37 names_2010-06-20.txt
-rw---xr-- 1 jovyan root 615 Nov  7 06:37 names_2012-07-10.dat
-rw---xr-- 1 jovyan root 652 Nov  7 06:37 names_2012-08-11.txt
-rw---xr-- 1 jovyan root 637 Nov  7 06:37 names_2012-08-31.txt
-rw---xr-- 1 jovyan root 638 Nov  7 06:37 names_2013-06-11.dat
-rw---xr-- 1 jovyan root 637 Nov  7 06:37 names_2013-10-31.txt
-rw---xr-- 1 jovyan root 644 Nov  7 06:37 names_2014-01-19.txt
-rw---xr-- 1 jovyan root 608 Nov  7 06:37 names_2016-11-29.dat



Podemos ver que, en efecto, los permisos fueron asignados correctamente.

___
__11.__ Cambia los permisos (por el método octal) de todos los archivos con terminación __.txt__ dentro de la carpeta `CUAUHTEMOC/sectorG/`, de tal manera que el dueño tenga lectura, escritura y ejecución, sólo lectura para el grupo y ningún permiso para los demás.

In [25]:
chmod 740 CUAUHTEMOC/sectorG/*.txt # Asignamos permisos
ls -l CUAUHTEMOC/sectorG/*txt # Verificamos permisos

-rwxr----- 1 jovyan root 623 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2000-04-14.txt
-rwxr----- 1 jovyan root 626 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2001-10-21.txt
-rwxr----- 1 jovyan root 626 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2002-01-07.txt
-rwxr----- 1 jovyan root 658 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2002-06-03.txt
-rwxr----- 1 jovyan root 611 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2002-07-31.txt
-rwxr----- 1 jovyan root 608 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2002-12-03.txt
-rwxr----- 1 jovyan root 641 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2003-06-06.txt
-rwxr----- 1 jovyan root 609 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2005-02-27.txt
-rwxr----- 1 jovyan root 628 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2005-06-11.txt
-rwxr----- 1 jovyan root 640 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2006-08-15.txt
-rwxr----- 1 jovyan root 654 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2006-10-26.txt
-rwxr----- 1 jovyan root 633 Nov  7 06:37 CUAUHTEMOC/sectorG/names_2007-12-11.txt
-rwxr----- 1 jov

___
__12.__ Cambia los permisos de todos los archivos con terminación __.dat__ dentro de la carpeta `CUAUHTEMOC/sectorP/`, de tal manera que el dueño tenga todos los permisos, sólo ejecución para el grupo y sólo lectura  para los demás. Aplica un `ls -l` para ver el resultado y guarda eso en un archivo con nombre `permisos.dat`

In [26]:
chmod 714 CUAUHTEMOC/sectorP/*.dat # Asignamos permisos
ls -l CUAUHTEMOC/sectorP/*dat > permisos.dat # Verificamos permisos y agregamos al archivo permisos.dat

In [27]:
head permisos.dat

-rwx--xr-- 1 jovyan root 643 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2000-03-28.dat
-rwx--xr-- 1 jovyan root 613 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2000-06-06.dat
-rwx--xr-- 1 jovyan root 620 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2000-10-02.dat
-rwx--xr-- 1 jovyan root 596 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2000-11-24.dat
-rwx--xr-- 1 jovyan root 605 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2000-11-27.dat
-rwx--xr-- 1 jovyan root 660 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2001-02-27.dat
-rwx--xr-- 1 jovyan root 632 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2002-02-05.dat
-rwx--xr-- 1 jovyan root 612 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2002-08-11.dat
-rwx--xr-- 1 jovyan root 612 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2004-09-20.dat
-rwx--xr-- 1 jovyan root 638 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2006-09-19.dat



___
__13.__ Utiliza el método (númerico o simbólico) que quieras para cambiar los permisos de todos los archivos en cuyo nombre este la fecha de mayo (05) que están dentro de la carpetas `CUAUHTEMOC/sectorP/` y `CUAUHTEMOC/sectorO/`. Los permisos deberán quedar de la siguiente manera todos los permisos para el dueño y ningún permiso para el grupo y los demás usuarios.

In [28]:
chmod 700 CUAUHTEMOC/sectorP/*-05-*
chmod 700 CUAUHTEMOC/sectorO/*-05-*

In [29]:
ls -l CUAUHTEMOC/sectorP/*-05-*

-rwx------ 1 jovyan root 623 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2017-05-17.dat
-rwx------ 1 jovyan root 621 Nov  7 06:37 CUAUHTEMOC/sectorP/names_2018-05-09.txt



In [30]:
ls -l CUAUHTEMOC/sectorO/*-05-*

-rwx------ 1 jovyan root 618 Nov  7 06:37 CUAUHTEMOC/sectorO/names_2003-05-04.dat
-rwx------ 1 jovyan root 627 Nov  7 06:37 CUAUHTEMOC/sectorO/names_2009-05-06.dat



___
## find y grep

__14.__ ¿Cuántos archivos tienen una extensión de `.txt` y cuántos tienen una extension `.dat`?

In [31]:
find ../directorio/*/*/. -name '*.txt' | wc -l

2036



In [32]:
find ../directorio/*/*/* -name '*.dat' | wc -l

2121



2036 archivos tienen la extensión `.txt` y 2121 archivos tienen la extensión `.dat`

____
__15.__ ¿Cuántos veces aparece _tu nombre_ (sin apellido) en todos los archivos? 


In [33]:
grep -o Rodrigo ../directorio/*/*/* | wc -l

519



El nombre __Rodrigo__ aparece un total de 519 veces en todos los archivos.

____
__16.__ ¿En cuántos archivos __NO__ aparecen tus nombres?

In [34]:
grep -L Rodrigo ../directorio/*/*/* | wc -l

3668



El __3668__ archivos no aparece el nombre __Rodrigo__

____
__17.__ ¿Cuántos veces coincide tus nombres con un archivo que tenga el año de tu nacimiento?

In [35]:
find ../directorio/*/*/names_2000* -type f | grep -H 'Rodrigo' | wc -l

0



Al aplicar la búsqueda con el filtro del año en que nací (__2000__) a los archivos y aplicar la búsqueda dentro de los archivos filtrados mi nombre (__Rodrigo__) obtenemos que aparece 0 veces.

____
__18.__ ¿De cuántos directorios y subdirectorios está compuesta la base de datos?

In [36]:
find ../directorio/* -maxdepth 0 -type d | wc -l

16



In [37]:
find ../directorio/* -mindepth 1 -type d | wc -l

170



In [38]:
find ../directorio/ -mindepth 1 -type f | wc -l

4157



La base de datos está compuesta de __16__ directorios principales, __170__ subdirectorios y __4157__ archivos.

____
__19.__ ¿Cuántos archivos tienen un tamaño menor a 1150 bytes?

In [39]:
find ../directorio/ -type f -size -1150c | wc -l

4157



__Todos__ los 4157 archivos que se encuentran en la base tienen un tamaño menor a 1150 bytes.

____
__20.__ ¿Cuántos archivos coinciden con el año 2020 en su nombre y además pertencen al `sectorS`?

In [40]:
find ../directorio/*/sectorS/names_2020* | wc -l

8



Son __8__ archivos que pertenecen a algún subdirectorio `sectorS` y que coinciden con el año 2020 en su nombre

____
Finalmente, comprimimos el directorio __Resultados__ con el comando especificado en el archivo de la Práctica 2: `tar czfv resultado.tar.gz Resultado/`

In [54]:
cd ..
tar czfv resultado.tar.gz Resultado/

Resultado/
Resultado/v
Resultado/todosarchivos.txt
Resultado/CUAUHTEMOC/
Resultado/CUAUHTEMOC/sectorJ/
Resultado/CUAUHTEMOC/sectorJ/names_2002-10-12.txt
Resultado/CUAUHTEMOC/sectorJ/names_2010-05-18.dat
Resultado/CUAUHTEMOC/sectorJ/names_2002-01-16.txt
Resultado/CUAUHTEMOC/sectorJ/names_2000-05-20.txt
Resultado/CUAUHTEMOC/sectorJ/names_2002-03-25.txt
Resultado/CUAUHTEMOC/sectorJ/names_2005-08-24.dat
Resultado/CUAUHTEMOC/sectorJ/names_2007-03-13.txt
Resultado/CUAUHTEMOC/sectorJ/names_2005-10-03.dat
Resultado/CUAUHTEMOC/sectorJ/names_2007-06-04.dat
Resultado/CUAUHTEMOC/sectorJ/names_2013-05-22.dat
Resultado/CUAUHTEMOC/sectorJ/names_2018-05-23.dat
Resultado/CUAUHTEMOC/sectorJ/names_2002-12-04.txt
Resultado/CUAUHTEMOC/sectorJ/names_2009-10-03.txt
Resultado/CUAUHTEMOC/sectorJ/names_2011-04-12.dat
Resultado/CUAUHTEMOC/sectorJ/names_2003-07-18.dat
Resultado/CUAUHTEMOC/sectorJ/names_2009-04-11.dat
Resultado/CUAUHTEMOC/sectorJ/names_2015-10-30.txt
Resultado/CUAUHTEMOC/sectorJ/names_2001-08-26.t